### PR TITLE
Feat/support more DBs

### PR DESCRIPTION
**Pull request by fengkiej**
**Before submitting this PR, please make sure**:
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/", etc)

**Optional**:
- Related Github issue: #1596 
- Related Clubhouse Story:

**A description of the changes proposed in the pull request**:
- The PR doesn't introduce breaking changes, so it can be tested with current test environment.
- Tested and running with Postgres Google SQL.
- Except for `sqlite` (default) option, a corresponding DBAPI need to be installed in Conda environment (ref: https://docs.sqlalchemy.org/en/13/dialects/) e.g. To use Postgres, psycopg2 need to be installed separately, we may consider to add DBAPIs in `setup/environment.yml` file.
- Config for DB options are available in `conf/conf_global.yml`.
- Unit test not yet available (any ideas how to test this?).
- After db connection initialized, it will remove all constraint on foreign key, due to sometimes, that TradeFill event may come before OrderCreated event.
- Documentations need to be added.

### CODE REVIEW

**Comment by martinkou** [comment]
Have you tested this on some of the databases? (e.g. postgresql)? It looks like to me that you'd need to at least include some of the driver modules (e.g. psycopg2cffi)

**Comment by fengkiej** [comment]
> Have you tested this on some of the databases? (e.g. postgresql)? It looks like to me that you'd need to at least include some of the driver modules (e.g. psycopg2cffi)

We actually running this version on prod.

That's correct external driver modules need to be added, we can include it into `environment.yml` or just make the feature optional for advanced user (with tutorial on documentation)

File: hummingbot/model/sql_connection_manager.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 from os.path import join
 from sqlalchemy import (
     create_engine,
+    inspect
```
Comment (martinkou) [inline hummingbot/model/sql_connection_manager.py:8]: incorrect syntax.

**Comment by martinkou** [comment]
Yes - I agree this is a feature for advanced users anyway - and thus it'll be on them to install the required database driver modules.

We've recently had an update to the config files - so you're gonna need to bump the template version by 1. I'll approve after that's done, and you've resolved the merge conflict and the missing comma bug identified above.

File: hummingbot/model/sql_connection_manager.py (hunk 1 above)
Comment (fengkiej) [reply hummingbot/model/sql_connection_manager.py:8]: thanks for noticing this part @martinkou

**Comment by martinkou** [comment]
retest this please

**Comment by martinkou** [comment]
It's failing unit tests at the moment. Still need some fixing.

```
======================================================================
ERROR: test_global_config_prompt_exists (test.test_config_templates.ConfigTemplatesUnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_config_templates.py", line 66, in test_global_config_prompt_exists
    if cvar.required:
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/client/config/config_var.py", line 45, in required
    return self._required_if()
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/client/config/global_config_map.py", line 318, in <lambda>
    required_if=lambda: any([x in global_config_map.get("db_engine").value for x in ["oracle", "postgres", "mysql", "mssql"]]),
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/client/config/global_config_map.py", line 318, in <listcomp>
    required_if=lambda: any([x in global_config_map.get("db_engine").value for x in ["oracle", "postgres", "mysql", "mssql"]]),
TypeError: argument of type 'NoneType' is not iterable
```

**Comment by fengkiej** [comment]
updated

**Comment by fengkiej** [comment]
added docs for feature

**Comment by martinkou** [comment]
retest this please

**Comment by martinkou** [comment]
It is still failing more test cases.

```

======================================================================
ERROR: Test performance analysis on a one exchange balance.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_performance_analysis.py", line 74, in test_basic_one_ex
    performance_analysis = PerformanceAnalysis()
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/client/performance_analysis.py", line 25, in __init__
    self.sql_manager = SQLConnectionManager.get_trade_fills_instance()
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 65, in get_trade_fills_instance
    cls._scm_trade_fills_instance = SQLConnectionManager(SQLConnectionType.TRADE_FILLS)
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 102, in __init__
    self._engine: Engine = self.get_db_engine(
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 72, in get_db_engine
    if "sqlite" in dialect:
TypeError: argument of type 'NoneType' is not iterable
-------------------- >> begin captured logging << --------------------
hummingbot.core.utils.async_utils: ERROR: Unhandled error in background task: 
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/data_feed/data_feed_base.py", line 61, in stop_network
    raise NotImplementedError
NotImplementedError
hummingbot.core.utils.async_utils: ERROR: Unhandled error in background task: 
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/data_feed/data_feed_base.py", line 61, in stop_network
    raise NotImplementedError
NotImplementedError
hummingbot.core.utils.exchange_rate_conversion: DEBUG: Waiting for coin_alpha_feed to get ready.
hummingbot.core.utils.exchange_rate_conversion: WARNING: Error initializing data feed - coin_alpha_feed.
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Test performance analysis on a two exchange balance with the same currencies trading in both exchanges.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_performance_analysis.py", line 112, in test_basic_two_ex
    performance_analysis = PerformanceAnalysis()
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/client/performance_analysis.py", line 25, in __init__
    self.sql_manager = SQLConnectionManager.get_trade_fills_instance()
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 65, in get_trade_fills_instance
    cls._scm_trade_fills_instance = SQLConnectionManager(SQLConnectionType.TRADE_FILLS)
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 102, in __init__
    self._engine: Engine = self.get_db_engine(
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 72, in get_db_engine
    if "sqlite" in dialect:
TypeError: argument of type 'NoneType' is not iterable

======================================================================
ERROR: Test performance analysis on a two exchange balance with different currencies trading. Note that this test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_performance_analysis.py", line 164, in test_different_tokens_two_ex
    performance_analysis = PerformanceAnalysis()
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/client/performance_analysis.py", line 25, in __init__
    self.sql_manager = SQLConnectionManager.get_trade_fills_instance()
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 65, in get_trade_fills_instance
    cls._scm_trade_fills_instance = SQLConnectionManager(SQLConnectionType.TRADE_FILLS)
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 102, in __init__
    self._engine: Engine = self.get_db_engine(
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 72, in get_db_engine
    if "sqlite" in dialect:
TypeError: argument of type 'NoneType' is not iterable

======================================================================
ERROR: Test the case where the starting balance is 0.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_performance_analysis.py", line 216, in test_nan_starting
    performance_analysis = PerformanceAnalysis()
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/client/performance_analysis.py", line 25, in __init__
    self.sql_manager = SQLConnectionManager.get_trade_fills_instance()
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 65, in get_trade_fills_instance
    cls._scm_trade_fills_instance = SQLConnectionManager(SQLConnectionType.TRADE_FILLS)
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 102, in __init__
    self._engine: Engine = self.get_db_engine(
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 72, in get_db_engine
    if "sqlite" in dialect:
TypeError: argument of type 'NoneType' is not iterable

======================================================================
ERROR: test suite for <class 'test.test_trade_performance_analysis.TestTradePerformanceAnalysis'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/miniconda/envs/hummingbot/lib/python3.8/site-packages/nose/suite.py", line 209, in run
    self.setUp()
  File "/root/miniconda/envs/hummingbot/lib/python3.8/site-packages/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/root/miniconda/envs/hummingbot/lib/python3.8/site-packages/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/root/miniconda/envs/hummingbot/lib/python3.8/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_trade_performance_analysis.py", line 104, in setUpClass
    cls.trade_fill_sql = SQLConnectionManager(SQLConnectionType.TRADE_FILLS, db_path="")
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 102, in __init__
    self._engine: Engine = self.get_db_engine(
  File "/var/jenkins_home/workspace/hb-test_pull-request/hummingbot/model/sql_connection_manager.py", line 72, in get_db_engine
    if "sqlite" in dialect:
TypeError: argument of type 'NoneType' is not iterable
```

Can you try to run
```
$ nosetests -v -d test/test*.py
```

locally and make sure everything works?

**Comment by fengkiej** [comment]
added fallback if `global_config_map.get("db_engine").value is none`, ran test with `pytest test/test*.py`, found 5 seemingly unrelated failures @martinkou 

```
================================================================================================================== FAILURES ===================================================================================================================
________________________________________________________________________________________ ConfigSecurityExistingPasswordUnitTest.test_existing_password ________________________________________________________________________________________

self = <test.test_config_security.ConfigSecurityExistingPasswordUnitTest testMethod=test_existing_password>

    def test_existing_password(self):
        loop = asyncio.get_event_loop()
>       loop.run_until_complete(self._test_existing_password())

test/test_config_security.py:75: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../miniconda3/envs/hummingbot/lib/python3.8/asyncio/base_events.py:616: in run_until_complete
    return future.result()
test/test_config_security.py:65: in _test_existing_password
    self.assertFalse(Security.is_decryption_done())
E   AssertionError: True is not false
______________________________________________________________________________________________ ExchangePriceManagerUnitTest.test_all_price_feeds ______________________________________________________________________________________________

self = <test.test_exchange_price_manager.ExchangePriceManagerUnitTest testMethod=test_all_price_feeds>

    def test_all_price_feeds(self):
        ev_loop = asyncio.get_event_loop()
        for ex_name in ExchangePriceManager.supported_exchanges:
            ExchangePriceManager.set_exchanges_to_feed([ex_name])
            ExchangePriceManager.start()
            ev_loop.run_until_complete(ExchangePriceManager.wait_til_ready())
            self.assertTrue(len(ExchangePriceManager.ex_feeds[ex_name].price_dict) > 0)
            for trading_pair, price in ExchangePriceManager.ex_feeds[ex_name].price_dict.items():
>               self.assertTrue("-" in trading_pair)
E               TypeError: argument of type 'NoneType' is not iterable

test/test_exchange_price_manager.py:22: TypeError
_________________________________________________________________________________________ ExchangePriceManagerMockAPIUnitTest.test_binance_mid_price __________________________________________________________________________________________

self = <test.test_exchange_price_manager.ExchangePriceManagerMockAPIUnitTest testMethod=test_binance_mid_price>

    def test_binance_mid_price(self):
        ExchangePriceManager.set_exchanges_to_feed(['binance'], False)
        ExchangePriceManager.start()
        self.ev_loop.run_until_complete(ExchangePriceManager.wait_til_ready())
        expected = (Decimal(FixtureExchangePrices.BINANCE[0]["bidPrice"]) +
                    Decimal(FixtureExchangePrices.BINANCE[0]["askPrice"])) / Decimal("2")
        eth_btc_price = ExchangePriceManager.get_price("ETH", "BTC")
>       self.assertEqual(expected, eth_btc_price)
E       AssertionError: Decimal('0.025') != Decimal('0.02496450')

test/test_exchange_price_manager.py:95: AssertionError
____________________________________________________________________________________________ ExchangePriceManagerMockAPIUnitTest.test_price_feeds _____________________________________________________________________________________________

self = <test.test_exchange_price_manager.ExchangePriceManagerMockAPIUnitTest testMethod=test_price_feeds>

    def test_price_feeds(self):
        ExchangePriceManager.set_exchanges_to_feed(['binance', 'liquid'])
        ExchangePriceManager.start()
        self.ev_loop.run_until_complete(ExchangePriceManager.wait_til_ready())
        self.assertTrue(BinancePriceFeed.get_instance()._ready_event.is_set())
>       self.assertEqual(len(BinancePriceFeed.get_instance().price_dict), 3)
E       AssertionError: 779 != 3

test/test_exchange_price_manager.py:56: AssertionError
__________________________________________________________________________________ ExchangePriceManagerMockAPIUnitTest.test_use_binance_when_none_supported ___________________________________________________________________________________

self = <test.test_exchange_price_manager.ExchangePriceManagerMockAPIUnitTest testMethod=test_use_binance_when_none_supported>

    def test_use_binance_when_none_supported(self):
        ExchangePriceManager.set_exchanges_to_feed(['coinbase_pro'])
        ExchangePriceManager.start()
        self.ev_loop.run_until_complete(ExchangePriceManager.wait_til_ready())
        self.assertTrue(BinancePriceFeed.get_instance()._ready_event.is_set())
>       self.assertEqual(len(BinancePriceFeed.get_instance().price_dict), 3)
E       AssertionError: 779 != 3

test/test_exchange_price_manager.py:75: AssertionError
```

**Comment by yzhang1994** [comment]
ok to test

**Pull request merged**